pkg/controllers/hpaautolabelretain/hpa_auto_label_retain_controller.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2023 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hpaautolabelretain
+
+import (
+	"context"
+
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/dynamic"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+const (
+	// ControllerName is the controller name that will be used when reporting events.
+	ControllerName = "hpa-auto-label-retain"
+	// scaleRefWorkerNum is the async Worker number
+	scaleRefWorkerNum = 1
+)
+
+// HPAAutoLabelRetain is to automatically add `retain-replicas` label to resource template mananged by HPA.
+type HPAAutoLabelRetain struct {
+	DynamicClient dynamic.Interface
```

Comment:
We can use `clientset.Interface` instead.

Reply:
ditto